[llvm] optionally install backend headers

arromanoff opened the pull request:
Currently, LLVM provides no way to access target-specific headers out-of-tree. This makes it practically impossible to use LLVM for instructions descriptions outside llvm-project repository and forces people to clone llvm and build all of it alongside their project. Or copy LLVM .tb files and run llvm-tblgen on them by hand. This commit adds a cmake option to install those headers alongside other llvm headers which makes it possible to use llvm backend headers out of llvm-project tree.

arrv-sc commented:
The story and reasoning behind this commit: https://discourse.llvm.org/t/installing-target-specific-headers-for-out-of-tree-usage/83080

Commit:
[llvm] optionally install backend headers
Currently, LLVM provides no way to access target-specific headers
out-of-tree. This makes it practically impossible to use LLVM for
instructions descriptions outside llvm-project repository and forces
people to clone llvm and build all of it alongside their project. Or
copy LLVM .tb files and run llvm-tblgen on them by hand. This commit
adds a cmake option to install those headers alongside other llvm
headers which makes it possible to use llvm backend headers out of
llvm-project tree.

Co-authored-by: Sergei Zimmerman <sergei.zimmerman@syntacore.com>

llvm-beanz commented:
I'm not really a fan of complicating our build by supporting this configuration, but that isn't a strongly held opinion.

Looping in some other contributors with build expertise: @petrhosek & @compnerd.

compnerd reviewed (commented):
I'm not sure how I feel about this change. I can see the value for out-of-tree clients. However, there are no ABI guarantees and there is no library versioning. As long as these are paired with a static library it _might_ be okay.

How do we ensure that the transitive closure is captured and nothing accidentally is left out?
